scripts/exglobal_diag.sh

```diff
@@ -170,113 +170,126 @@ EOFdiag
+                echo "diag_${type}_${string}.${CDATE}*" >> "${diaglist[n]}" 
+                numfile[n]=$(expr "${numfile[n]}" + 1)
             fi
          done
       done
-      echo $(date) END loop $string >&2
+      echo "$(date)" END loop "${string}" >&2 || true
    done
 
    # We should already be in $DATA, but extra cd to be sure.
-   cd $DATA
+   cd "${DATA}" || exit
 
    # If requested, compress diagnostic files
-   if [ $DIAG_COMPRESS = "YES" -a $USE_CFP = "NO" ]; then
-      echo $(date) START $COMPRESS diagnostic files >&2
-      for file in $(ls diag_*${CDATE}${DIAG_SUFFIX}); do
-         $COMPRESS $file
+   if [[ "${DIAG_COMPRESS}" = "YES" && "${USE_CFP}" = "NO" ]]; then
+      echo "$(date)" START "${COMPRESS}" diagnostic files >&2 || true
+      # shellcheck disable=SC2045 
+      for file in $(ls "diag_*${CDATE}${DIAG_SUFFIX}"); do
+         "${COMPRESS}" "${file}"
       done
-      echo $(date) END $COMPRESS diagnostic files >&2
+      echo "$(date)" END "${COMPRESS}" diagnostic files >&2 || true
    fi
 
-   if [ $USE_CFP = "YES" ] ; then
-      chmod 755 $DATA/mp_diag.sh
-      ncmd=$(cat $DATA/mp_diag.sh | wc -l)
-      if [ $ncmd -gt 0 ]; then
+   if [[ "${USE_CFP}" = "YES" ]] ; then
+      chmod 755 "${DATA}/mp_diag.sh"
+      # shellcheck disable=SC2002,SC2312 
+      ncmd=$(cat "${DATA}/mp_diag.sh" | wc -l)
+      # shellcheck disable=SC2034
+      if [[ "${ncmd}" -gt 0 ]]; then
          ncmd_max=$((ncmd < npe_node_max ? ncmd : npe_node_max))
+         # shellcheck disable=SC2086,SC2250 
          APRUNCFP_DIAG=$(eval echo $APRUNCFP)
-         $APRUNCFP_DIAG $DATA/mp_diag.sh
+         "${APRUNCFP_DIAG}" "${DATA}/mp_diag.sh"
          export err=$?; err_chk
       fi
    fi
 
    # Restrict diagnostic files containing rstprod data
    rlist="conv_gps conv_ps conv_pw conv_q conv_sst conv_t conv_uv saphir"
-   for rtype in $rlist; do
+   for rtype in ${rlist}; do
       set +e
-      ${CHGRP_CMD} *${rtype}*
-      ${STRICT_ON:-set -e}
+      # shellcheck disable=SC2045,SC2035,SC2086 
+      for filename in $(ls *${rtype}*); do
+         if [[ -s "${filename}" ]]; then
+            "${CHGRP_CMD}" "${filename}"
+            ${STRICT_ON:-set -e}
+         fi
+      done
    done
 
    # If requested, create diagnostic file tarballs
-   if [ $DIAG_TARBALL = "YES" ]; then
-      echo $(date) START tar diagnostic files >&2
+   if [[ "${DIAG_TARBALL}" = "YES" ]]; then
+      echo "$(date)" START tar diagnostic files >&2 || true
       n=-1
-      while [ $((n+=1)) -le $ntype ] ;do
+      while [[ $((n+=1)) -le "${ntype}" ]] ;do
          TAROPTS="-uvf"
-         if [ ! -s ${diagfile[n]} ]; then
+         if [[ ! -s "${diagfile[n]}" ]]; then
             TAROPTS="-cvf"
          fi
-         if [ ${numfile[n]} -gt 0 ]; then
-            tar $TAROPTS ${diagfile[n]} $(cat ${diaglist[n]})
+         if [[ "${numfile[n]}" -gt 0 ]]; then
+            # shellcheck disable=SC2046
```

Comment:
Depending on which part of the next line it is complaining about, we may need to keep this (with a note added about what it is we are ignoring). Try removing it, then add it back if there are still complaints.